src/libos/crates/io-uring-callback/src/io_handle.rs

```diff
@@ -87,8 +94,8 @@ pub(crate) struct IoToken {
 }
```

Comment:
Canceling is different from normal I/O in that
1. It does not make sense to hold a handle for cancel;
2. It does not make sense to maintain state for cancel;
3. It does not make sense to add callback (the callback you write is essentially empty) for cancel.

For the reasons above, I think we should extend the notion of `IoToken`, with which we can differentiate between normal I/O operations and cancel operations.
```rust
pub enum IoToken {
    Request(IoRequest),
    Cancel,
}

pub struct IoRequest {
    state: IoState,
    callback: Option<Callback>,
    waker: Option<Waker>,
}

pub struct IoHandle(Arc<IoRequest>);
```

Reply:
I agree.

Canceling in io-uring is different from other operations: **We don't have to wait for it to complete**.

- If canceling successes later, it's ok.
- If canceling fails later, it's still ok, since the operation itself may not support cancellation or has been completed.

Reply:
Another point, we can only record IoRequests, implement canceling only by a inner interface (no token).

maybe we need two cancel interface?

- non-block cancel: just send cancel request, no wait
- block cancel: send and wait (await)